### PR TITLE
add postgres persistence v2

### PR DESCRIPTION
**Pull request by mactaggart**
This PR adds postgres persistence. This PR provides an independent postgres persistence module leaving the existing mysql persistence module unmodified and introduces no breaking changes. (eg. mysql properties remain the same) This PR supersedes [1327](https://github.com/Netflix/conductor/pull/1327) which introduced a sql-persisence module that contained code common to both mysql-persistence and postgres-persistence.

### CODE REVIEW

File: postgres-persistence/build.gradle [hunk 1]
```diff
@@ -0,0 +1,19 @@
+dependencies {
+    compile project(':conductor-core')
+    compile "com.google.inject:guice:${revGuice}"
+
+    compile "commons-io:commons-io:${revCommonsIo}"
+    compile "org.postgresql:postgresql:42.2.6"
```
Comment (apanicker-nflx) [inline postgres-persistence/build.gradle]: Please move version to `versionsOfDependencies.gradle`

Comment (apanicker-nflx) [inline postgres-persistence/build.gradle]: Please move version to `versionsOfDependencies.gradle`

File: postgres-persistence/build.gradle [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    compile "com.zaxxer:HikariCP:${revHikariCP}"
+    compile "org.flywaydb:flyway-core:${revFlywayCore}"
+
+    testCompile project(':conductor-core').sourceSets.test.output
+    testCompile "org.slf4j:slf4j-simple:1.7.25"
+    testCompile 'com.opentable.components:otj-pg-embedded:0.13.1'
```
Comment (apanicker-nflx) [inline postgres-persistence/build.gradle]: Same as above.

File: postgres-persistence/src/main/java/com/netflix/conductor/dao/postgres/PostgresBaseDAO.java [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+import java.util.List;
+import java.util.function.Consumer;
+
+import static com.netflix.conductor.core.execution.ApplicationException.Code.*;
+import static java.lang.Integer.parseInt;
+import static java.lang.System.getProperty;
+
+public abstract class PostgresBaseDAO {
+
+    private static final int ER_LOCK_DEADLOCK = -1; //FIXME; replace with error code postgres returns for deadlock
```
Comment (apanicker-nflx) [inline postgres-persistence/src/main/java/com/netflix/conductor/dao/postgres/PostgresBaseDAO.java]: Is this still to be fixed?

File: postgres-persistence/src/main/java/com/netflix/conductor/dao/postgres/PostgresMetadataDAO.java [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+import com.netflix.conductor.core.config.Configuration;
+import com.netflix.conductor.core.execution.ApplicationException;
+import com.netflix.conductor.dao.MetadataDAO;
+import com.netflix.conductor.metrics.Monitors;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.util.*;
```
Comment (apanicker-nflx) [inline postgres-persistence/src/main/java/com/netflix/conductor/dao/postgres/PostgresMetadataDAO.java]: Please avoid wildcard imports

File: postgres-persistence/build.gradle (hunk 1 above)
Comment (mactaggart) [reply postgres-persistence/build.gradle]: ok; updated

File: postgres-persistence/build.gradle (hunk 2 above)
Comment (mactaggart) [reply postgres-persistence/build.gradle]: updated

File: postgres-persistence/src/main/java/com/netflix/conductor/dao/postgres/PostgresMetadataDAO.java (hunk 4 above)
Comment (mactaggart) [reply postgres-persistence/src/main/java/com/netflix/conductor/dao/postgres/PostgresMetadataDAO.java]: updated to remove all wildcard imports

**Comment by mactaggart** [comment]
Added license to all java files

**Comment by apanicker-nflx** [comment]
@mactaggart Thank you for making the changes. Please resolve conflicts and we should be good to merge.

**Comment by mactaggart** [comment]
@apanicker-nflx I resolved conflicts and squashed commits.

**Comment by checkingchecks** [comment]
Hi, I encountered the following error:

`FlywayException: Found non-empty schema "cndr_schema" without metadata table! Use baseline() or set baselineOnMigrate to true to initialize the metadata table.`

I put `flyway.baselineOnMigrate=true` in my properties which I include when running the server but it seems it doesn't work cos I still get the same error.

**Comment by mactaggart** [comment]
@checkingchecks I'll look into this

**Comment by checkingchecks** [comment]
> @checkingchecks I'll look into this

hi, any updates? so I initially created the same schema and tables via script to set it all up before running the server but when I ran the server it still prompts the flyway migration exception even though it has the similar tables and indexes before running.

**Comment by checkingchecks** [comment]
> @checkingchecks I'll look into this

hi, any updates? so I initially created the same schema and tables via script to set it all up before running the server but when I ran the server it still prompts the flyway migration exception even though it has the similar tables and indexes before running.

**Comment by mactaggart** [comment]
@checkingchecks While flyaway has an option for baselineOnMigrate, conductor does not. I created a branch that adds this configuration option to conductor. To test this, create a new build from my branch:

```
git clone -b baseline-on-migrate https://github.com/mactaggart/conductor.git
cd conductor
./gradlew build -x test
```

In your conductor config, set:

`baseline.on.migrate=true`

If this works in your environment, I can submit a PR. Hope this helps.

**Comment by checkingchecks** [comment]
> @checkingchecks While flyaway has an option for baselineOnMigrate, conductor does not. I created a branch that adds this configuration option to conductor. To test this, create a new build from my branch:
> 
> ```
> git clone -b baseline-on-migrate https://github.com/mactaggart/conductor.git
> cd conductor
> ./gradlew build -x test
> ```
> 
> In your conductor config, set:
> 
> `baseline.on.migrate=true`
> 
> If this works in your environment, I can submit a PR. Hope this helps.

thanks. i'm trying it out with conductor server 2.24.5 but i think the changes are not reflected when applying `baseline.on.migrate=true`. Is there a part of the config that i haven't done?

**Comment by mactaggart** [comment]
do you also have `flyaway.enabled=true`?

**Comment by mcancio** [comment]
> Hi, I encountered the following error:
> 
> `FlywayException: Found non-empty schema "conductor_schema" without metadata table! Use baseline() or set baselineOnMigrate to true to initialize the metadata table.`
> 
> I put `flyway.baselineOnMigrate=true` in my properties which I include when running the server but it seems it doesn't work cos I still get the same error.

Hi @checkingchecks, I had this issue when the Postgres schema I was using was already pre-populated with tables. To work around the the FlywayException issue,
1. Configured Conductor an empty schema, "new_conductor_schema".
2. Start Conductor - this used flyway to create the objects.
3. (optional) migrated the data from the old schema into the new one.

The succeeding runs of Conductor no longer had FlywayExceptions.

**Pull request merged**